README.rst
watchman
========

Watch directories for new files and send tasks to a queue for further
asynchronous processing. This project is deliberately lightweight
(simple configuration via files, no database and/or web server).
We use watchman on our FTP server to automatically process
uploaded files.


Building watchman
-----------------

Development with Vagrant, virtualenv, etc::

    $ git clone git@github.com:nens/watchman.git
    $ cd watchman
    $ ln -s profiles/development.cfg buildout.cfg
    $ python bootstrap.py
    $ bin/buildout

Development with Docker::

    $ git clone git@github.com:nens/watchman.git
    $ cd watchman
    $ ln -s profiles/development.cfg buildout.cfg
    $ docker-compose build
    $ docker volume create --name=ftproot
    $ docker-compose run --rm watchman python bootstrap.py
    $ docker-compose run --rm watchman bin/buildout

NB: We share the volume `ftproot` with another Docker project. Adjust to your
needs.


Configuring watchman
--------------------

Firstly, create a celeryconfig.py and configure Celery as desired,
for example::

    CELERYD_HIJACK_ROOT_LOGGER = False

By default, tasks are sent to a queue named "celery". If a task needs to be
routed to a specific queue, this may be done as follows::

    CELERY_ROUTES = {
        'lizard_nxt.tasks.import_raster_task': {'queue': 'single_worker'},
    }

NB: Celery v4 uses new lowercase settings. Check out the documentation.

Since multiple brokers are supported, do not define a `BROKER_URL` here as you
would typically do when working with Celery. (NB: This setting supports a
list of broker URLs, but this is for use in a failover strategy.)

Secondly, copy config-example.py to config.py and configure logging and
pyinotify.

If watchman is used to recursively watch directories created by
ProFTPD, use this setting::

    ProFTPD = True

Without this setting, Pyinotify's feature to recursively watch
newly created directories doesn't function properly.

Configure one or more brokers::

    BROKERS = {
        'lizard': 'amqp://guest:guest@rabbitmq:5672//',
        '3di': 'redis://redis:6379/0',
    }

WATCHES lists the directories to monitor for new files::

    WATCHES = [
        {'path': '/srv/ftp', 'rec': True, 'auto_add': True},
    ]

See WatchManager's add_watch function: http://seb-m.github.io/pyinotify/.

PATTERNS maps file extensions on Celery tasks::

    PATTERNS = {
        ('/srv/ftp/*.csv', {'broker': 'lizard', 'task': 'lizard_task'}),
        ('/srv/ftp/*.nc', {'broker': '3di', 'task': '3di_task'}),
    }

See fnmatch: https://docs.python.org/2/library/fnmatch.html.

Sometimes order matters (most specific first)::

    PATTERNS = OrderedDict([
        ('/srv/ftp/*/events/*.csv', {'broker': 'lizard', 'task': 'process_events'}),
        ('/srv/ftp/*.csv', {'broker': 'lizard', 'task': 'process_timeseries'}),
    ])

New files may be moved to another location (e.g. permanent storage) before
submitting tasks. This is optional (you may leave this dictionary empty)::

    MOVES = {
        '/srv/ftp': '/isilon/foo/bar',
    }

This will move ``/srv/ftp/username/lizard/events.csv`` to
``/isilon/foo/bar/username/lizard/events.csv``
before sending the task.

Running watchman
----------------

Interactively::

    $ bin/notify

Daemonized::

    $ bin/supervisord

Development with docker::

    $ docker-compose up

Verify that it works::

    $ docker exec -it watchman_watchman_1 /bin/bash
    $ cd /srv/ftp
    $ touch data.csv

The log window should display something like this::

    watchman_1  | 2016-12-28 16:17:13,803 [INFO] watchman.notify: Notified of /srv/ftp/data.csv
    watchman_1  | 2016-12-28 16:17:13,827 [INFO] watchman.notify: Sending task lizard_task

Visit the RabbitMQ management interface in your browser and inspect the celery queue::

    http://localhost:15672/#/queues


CHANGES.rst
Changelog of watchman
===================================================


0.1 (unreleased)
----------------

- Initial project structure created with nensskel 1.35.

- Match on pathname instead of filename.

- Auto creating 'accepted' and 'rejected' folders.

- Match relevant library versions with those of lizard-nxt to prevent this
  issue https://github.com/celery/celery/issues/4356.

- Bumped celery to 4.3

- Fixed bug with undefined connection.

- Uniquify filenames when copying to long-term storage.

- Add a 'kwargs' key to the PATTERNS in the celery config. This enables setting
  custom priority, queue, expires, etc. for some path patterns.


watchman/notify.py
# (c) Nelen & Schuurmans.  GPL licensed, see LICENSE.rst.
# -*- coding: utf-8 -*-

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import fnmatch
import logging.config
import os
import shutil
import time

import pyinotify

from . import config as cfg
from .celery import CONNECTION_POOLS
from .celery import app

logger = logging.getLogger(__name__)


def mkdir(path):
    if not os.path.exists(path):
        logger.debug('Creating directory %s', path)
        os.makedirs(path)
    else:
        assert(os.path.isdir(path))


def move(src, dst):
    logger.info('Moving %s to %s', src, dst)
    shutil.move(src, dst)


def uniquify(filename):
    """Make a filename unique by adding milliseconds since epoch.

    For example: foo.txt => foo@1596705456400.txt
    """
    root, ext = os.path.splitext(filename)  # foo.txt => ('foo', '.txt')
    ms = time.time() * 1000  # milliseconds since epoch as a float
    return "{root}@{ms:.0f}{ext}".format(root=root, ms=ms, ext=ext)

class FileHandler(pyinotify.ProcessEvent):
    """An event handler for new/updated files.

    Depending on the file extension, a task is sent to a task queue for further
    asynchronous processing. Only one task per file is sent, even in case of
    multiple matches.

    """
    ProFTPD = getattr(cfg, 'ProFTPD', False)

    def my_init(self, watch_manager={}):
        # The recommended way is not to override __init__ in your
        # subclass, but to define a my_init() method instead.
        # See pyinotify.ProcessEvent.
        self.wm = watch_manager

    def process_default(self, event):
        # If we are not watching directories created by ProFTPD,
        # return immediately.
        if not FileHandler.ProFTPD:
            return
        # Directory creation by ProFTPD is a two-step process. First a hidden
        # directory is created, e.g. `.dstXXX59N0g9`. Then, this directory
        # is renamed. This screws up the `auto_add` feature of pyinotify,
        # which automatically adds watches on newly created directories:
        # by the time pyinotify tries to add a watch, the dot folder
        # doesn't exist anymore, leaving the directory unwatched.
        if event.mask == pyinotify.IN_MOVED_TO | pyinotify.IN_ISDIR:
            if os.path.basename(event.src_pathname).startswith(".dst"):
                watches = {w.path: w for w in self.wm.watches.values()}
                parent_dir = os.path.dirname(event.pathname)
                parent_watch = watches.get(parent_dir)
                if (parent_watch and parent_watch.auto_add
                        and event.pathname not in watches):
                    self.wm.add_watch(
                        event.pathname, parent_watch.mask, auto_add=True)

    def process_IN_CLOSE_WRITE(self, event):

        logger.info('Notified of %s', event.pathname)

        # Some setups require a file to be moved to a shared location first.

        # If k = '/srv/ftp' and v = '/isilon/prod/data', then
        # '/srv/ftp/username/lizard/events.csv' will be moved to
        # '/isilon/prod/data/username/lizard/events@1596699866190.csv'.
        # The @milliseconds part is added to prevent filename clashes.

        pathname = event.pathname
        for k, v in cfg.MOVES.items():
            if event.pathname.startswith(k):
                src = event.pathname
                sub = event.path[len(k):].lstrip('/')
                dst = os.path.join(v, sub)
                try:
                    mkdir(dst)
                    mkdir(os.path.join(dst, 'accepted'))
                    mkdir(os.path.join(dst, 'rejected'))
                    filename = uniquify(event.name)
                    pathname = os.path.join(dst, filename)
                    move(src, pathname)
                except Exception as e:
                    logger.error(e)
                    return
                break

        # Now that the file has been moved, notify the task server.

        for pattern, data in cfg.PATTERNS.iteritems():
            if fnmatch.fnmatch(event.pathname, pattern):
                connection = None
                try:
                    task = data['task']
                    broker = data['broker']
                    kwargs = data.get('kwargs', {})
                    pool = CONNECTION_POOLS[broker]
                    connection = pool.acquire()
                    logger.info('Sending task %s', task)
                    app.send_task(task, args=[pathname], connection=connection, **kwargs)
                except Exception as e:
                    logger.error(e)
                finally:
                    if connection is not None:
                        connection.release()
                    break


def get_notifier():
    """Link directories being watched with an event handler.

    When a file that was opened for writing is closed, a FileHandler is
    notified.

    """
    wm = pyinotify.WatchManager()
    mask = pyinotify.ALL_EVENTS
    for watch in cfg.WATCHES:
        wm.add_watch(mask=mask, **watch)
    notifier = pyinotify.Notifier(wm, FileHandler(watch_manager=wm))
    return notifier


def main():
    logging.config.dictConfig(cfg.LOGGING)
    notifier = get_notifier()
    notifier.loop()

if __name__ == '__main__':
    main()


watchman/config-example.py
# -*- coding: utf-8 -*-

from collections import OrderedDict
import os

SETTINGS_DIR = os.path.dirname(os.path.realpath(__file__))
BUILDOUT_DIR = os.path.abspath(os.path.join(SETTINGS_DIR, '..'))

LOGGING = {
    'version': 1,
    'formatters': {
        'verbose': {
            'format': '%(asctime)s [%(levelname)s] %(name)s: %(message)s'
        },
    },
    'handlers': {
        'console': {
            'class': 'logging.StreamHandler',
            'formatter': 'verbose',
            'level': 'DEBUG',
        },
        'file': {
            'class': 'logging.handlers.RotatingFileHandler',
            'filename': os.path.join(
                BUILDOUT_DIR, 'var', 'log', 'watchman.log'),
            'formatter': 'verbose',
            'level': 'INFO',
        },
    },
    'loggers': {
        '': {
            'handlers': ['console', 'file'],
            'level': 'DEBUG',
        },
        'watchman.notify': {
            'propagate': True,
        },
    },
}

ProFTPD = True

BROKERS = {
    '3di': 'redis://redis:6379/0',
    'lizard': 'amqp://guest:guest@rabbitmq:5672//',
}

WATCHES = [
    {'path': '/srv/ftp', 'rec': True, 'auto_add': True},
]

PATTERNS = OrderedDict([
    ('/srv/ftp/priority-user/*.csv', {'broker': 'lizard', 'task': 'lizard_task', 'kwargs': {'priority': 5}}),
    ('/srv/ftp/expiring-import/*.csv', {'broker': 'lizard', 'task': 'lizard_task', 'kwargs': {'expires': 300}}),
    ('/srv/ftp/*.csv', {'broker': 'lizard', 'task': 'lizard_task'}),
    ('/srv/ftp/*.nc', {'broker': '3di', 'task': '3di_task'}),
])

MOVES = {
    '/foo': '/isilon/foo',
    '/bar': '/isilon/bar',
}
